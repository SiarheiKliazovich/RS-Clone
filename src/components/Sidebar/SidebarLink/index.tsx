import { FunctionComponent, useState } from "react";
import { Link } from "react-router-dom";
import { SidebarIcon } from "../SidebarIcon";
import { SidebarLinkType } from "../../../types";

const SidebarLink: FunctionComponent<SidebarLinkType> = ({
  type,
  to,
  title,
  color,
}) => {
  const [isHovered, setIsHovered] = useState(false);

  return (
    <Link className="header__link" to={to}>
      <li
        className="sidebar__item"
        onMouseOver={() => setIsHovered(true)}
        onMouseOut={() => setIsHovered(false)}
      >
        <SidebarIcon
          color={isHovered ? color.hover : color.static}
          type={type}
        />
        <span
          style={{ color: isHovered ? color.hover : color.static }}
          className="sidebar__name"
        >
          {title}
        </span>
      </li>
    </Link>
  );
};

export default SidebarLink;
