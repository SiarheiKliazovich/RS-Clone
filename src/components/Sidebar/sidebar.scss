@import "../../sass/mixins";
@import "../../sass/variables";

.sidebar {
  height: calc(100vh - 80px);
  margin-top: 80px;
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 0;
  left: 0;
  opacity: 0;
  background-color: $color-element-background-light;
  transition: $transition;
  transform: translateX(-210px);
  box-shadow: $box-shadow;
  overflow: hidden;
  z-index: 5;

  &--fixed {
    opacity: 1;
    transform: translateX(0);

    .sidebar__name {
      display: none;
    }
  }

  &--show {
    width: 210px;
    opacity: 1;
    transform: translateX(0);

    .sidebar__name {
      display: inline;
    }
  }

  @include media-small-mobile() {
    height: calc(100vh - 70px);
    margin-top: 70px;
  }
}

.sidebar__wrapper {
  margin: 35px;

  @include media-small-mobile {
    margin: 35px 20px;
  }
}

.sidebar__list {
  display: flex;
  flex-direction: column;
  gap: 50px;
}

.sidebar__item {
  display: flex;
  align-items: center;
  gap: 25px;
  cursor: pointer;

  &:hover {
    .sidebar__name {
      color: $color-accent-hover;
    }

    & > svg {
      stroke: $color-accent-hover;
    }
  }
}

.sidebar__icon {
  width: 25px;
  height: 25px;
  min-width: 25px;
  min-height: 25px;
  display: inline-block;
  transition: $transition;
}

.sidebar__name {
  font-size: $font-size-lg;
  font-weight: $font-weight-bold;
}
