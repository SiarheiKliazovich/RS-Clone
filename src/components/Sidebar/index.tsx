import "./sidebar.scss";
import { FunctionComponent } from "react";
import { Link } from "react-router-dom";
import { SidebarType } from "../../types";
import classNames from "classnames";
import { SidebarIcon } from "./SidebarIcon";
import { lightTheme, nightTheme } from "../../data/constants";

const Sidebar: FunctionComponent<SidebarType> = ({
  accentColor,
  isSidebarFixed,
  isSidebarAccentMode,
  isSidebarShow,
  isNightMode,
}: SidebarType) => {
  const backgroundColor = isNightMode
    ? nightTheme.background.element
    : lightTheme.background.element;

  const itemColor = isSidebarAccentMode
    ? nightTheme.sidebarFontColor
    : accentColor.static;

  return (
    <nav
      className={classNames(
        "sidebar",
        { "sidebar--fixed": isSidebarFixed },
        { "sidebar--show": isSidebarShow }
      )}
      style={
        isSidebarAccentMode
          ? { backgroundColor: accentColor.static, color: itemColor }
          : { backgroundColor: backgroundColor, color: accentColor.static }
      }
    >
      <div className="sidebar__wrapper">
        <ul className="sidebar__list">
          <Link className="header__link" to="">
            <li className="sidebar__item">
              <SidebarIcon color={itemColor} type={"dashboard"} />
              <span className="sidebar__name">Dashboard</span>
            </li>
          </Link>
          <li className="sidebar__item">
            <SidebarIcon color={itemColor} type={"calendar"} />
            <span className="sidebar__name">Calendar</span>
          </li>
          <li className="sidebar__item">
            <SidebarIcon color={itemColor} type={"orders"} />
            <span className="sidebar__name">Orders</span>
          </li>
          <Link className="header__link" to="users">
            <li className="sidebar__item">
              <SidebarIcon color={itemColor} type={"users"} />
              <span className="sidebar__name">Users</span>
            </li>
          </Link>
        </ul>
      </div>
    </nav>
  );
};

export default Sidebar;
