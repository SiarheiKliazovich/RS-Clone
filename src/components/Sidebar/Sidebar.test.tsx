import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import Sidebar from ".";

describe("Sidebar", () => {
  test("renders correctly", () => {
    const { container } = render(
      <Router>
        <Sidebar
          accentColor={{ static: "#ff0000", hover: "#00ff00" }}
          isSidebarFixed={true}
          isSidebarAccentMode={true}
          isSidebarShow={true}
          isNightMode={false}
        />
      </Router>
    );
    expect(container).toMatchSnapshot();
  });
});
