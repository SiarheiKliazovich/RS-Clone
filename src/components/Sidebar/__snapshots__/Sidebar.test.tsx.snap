// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders correctly 1`] = `
<div>
  <nav
    class="sidebar sidebar--fixed sidebar--show"
    style="background-color: rgb(255, 0, 0); color: rgb(255, 255, 255);"
  >
    <div
      class="sidebar__wrapper"
    >
      <ul
        class="sidebar__list"
      >
        <a
          class="header__link"
          href="/"
        >
          <li
            class="sidebar__item"
          >
            <svg
              class="sidebar__icon"
              fill="none"
              stroke="#ffffff"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1"
              />
              <polygon
                points="12 15 17 21 7 21 12 15"
              />
            </svg>
            <span
              class="sidebar__name"
            >
              menu.dashboard
            </span>
          </li>
        </a>
        <li
          class="sidebar__item"
        >
          <svg
            class="sidebar__icon"
            fill="none"
            stroke="#ffffff"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="4"
            />
            <line
              x1="16"
              x2="16"
              y1="2"
              y2="6"
            />
            <line
              x1="8"
              x2="8"
              y1="2"
              y2="6"
            />
            <line
              x1="3"
              x2="21"
              y1="10"
              y2="10"
            />
          </svg>
          <span
            class="sidebar__name"
          >
            menu.calendar
          </span>
        </li>
        <a
          class="header__link"
          href="/orders"
        >
          <li
            class="sidebar__item"
          >
            <svg
              class="sidebar__icon"
              fill="none"
              stroke="#ffffff"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="9"
                cy="21"
                r="1"
              />
              <circle
                cx="20"
                cy="21"
                r="1"
              />
              <path
                d="M1 1h4l2.68 13.39a2 2 0 0 0 2 1.61h9.72a2 2 0 0 0 2-1.61L23 6H6"
              />
            </svg>
            <span
              class="sidebar__name"
            >
              menu.orders
            </span>
          </li>
        </a>
        <a
          class="header__link"
          href="/users"
        >
          <li
            class="sidebar__item"
          >
            <svg
              class="sidebar__icon"
              fill="none"
              stroke="#ffffff"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 21v-2a4 4 0 0 0-4-4H5a4 4 0 0 0-4 4v2"
              />
              <circle
                cx="9"
                cy="7"
                r="4"
              />
              <path
                d="M23 21v-2a4 4 0 0 0-3-3.87"
              />
              <path
                d="M16 3.13a4 4 0 0 1 0 7.75"
              />
            </svg>
            <span
              class="sidebar__name"
            >
              menu.users
            </span>
          </li>
        </a>
      </ul>
    </div>
  </nav>
</div>
`;
